pkg/apis/sources/v1alpha1/jobsource_lifecycle.go

```diff
@@ -21,25 +21,80 @@ import (
 	"knative.dev/pkg/apis"
 )
 
-var condSet = apis.NewLivingConditionSet()
+var condSet = apis.NewBatchConditionSet(
+	JobSourceConditionSinkProvided,
+	JobSourceConditionJobSucceeded,
+)
 
 // GetGroupVersionKind implements kmeta.OwnerRefable
 func (js *JobSource) GetGroupVersionKind() schema.GroupVersionKind {
 	return SchemeGroupVersion.WithKind("JobSource")
 }
 
-func (jss *JobSourceStatus) InitializeConditions() {
-	condSet.Manage(jss).InitializeConditions()
+func (s *JobSourceStatus) InitializeConditions() {
+	condSet.Manage(s).InitializeConditions()
+}
+
+// Succeeded returns true if the JobSource has succeeded.
+func (s *JobSourceStatus) Succeeded() bool {
+	// This condition set should use apis.ConditionSucceeded (JobSourceConditionSucceeded).
+	return condSet.Manage(s).IsHappy()
+}
+
+// MarkSink sets the conditions that the source has received a sink URI.
+func (s *JobSourceStatus) MarkSink(uri string) {
+	if s.IsRunning() {
+		// If the sink changes while the job is already running, we are choosing to let the job finish
+		// using the outdated sink.
+		// TODO(spencer-p) Log this somewhere.
+		// TODO(spencer-p,n3wscott) This is opinionated, discuss further. Other options include:
+		//  - Proxying the sink and resolving it only when the cloud event is generated
+		//  - Killing the job and restarting it
+		return
+	}
+
+	s.SinkURI = uri
+	if len(uri) > 0 {
+		condSet.Manage(s).MarkTrue(JobSourceConditionSinkProvided)
+	} else {
+		condSet.Manage(s).MarkUnknown(JobSourceConditionSinkProvided, "SinkEmpty", "Sink resolved to empty URI")
+	}
+}
+
+// MarkNoSink sets the condition that the JobSource does not have a sink configured.
+func (s *JobSourceStatus) MarkNoSink(reason, messageFormat string, messageA ...interface{}) {
+	condSet.Manage(s).MarkFalse(JobSourceConditionSinkProvided, reason, messageFormat, messageA...)
+}
+
+// JobSucceeded returns true if the underlying Job has succeeded.
+func (s *JobSourceStatus) JobSucceeded() bool {
+	return condSet.Manage(s).GetCondition(JobSourceConditionJobSucceeded).IsTrue()
+}
+
+// IsRunning returns true if the job is currently running.
+func (s *JobSourceStatus) IsRunning() bool {
+	jobsucceeded := condSet.Manage(s).GetCondition(JobSourceConditionJobSucceeded)
+	if !jobsucceeded.IsUnknown() {
```

Comment:
it might need more than this, like IsUnknown and Reason is "Running" because after you init and the conditions set to unknown the sink uri will be un-settable.

Reply:
ooh you do that later. 

Missread...